lib/jsdom/living/nodes/HTMLButtonElement-impl.js

```diff
@@ -47,6 +49,60 @@ class HTMLButtonElementImpl extends HTMLElementImpl {
         break;
     }
   }
+
+  get willValidate() {
+    return !isDisabled(this);
+  }
+
+  get validity() {
+    if (!this._validity) {
+      this._validity = ValidityState.createImpl({
+        willValidate: () => this.willValidate,
+        // Is a Boolean indicating the element's custom
+        // validity message has been set to a non-empty string by
+        // calling the element's setCustomValidity() method.
+        customError: () => this._customValidityMessage != null &&
+              this._customValidityMessage !== "",
+      });
+    }
+    return this._validity;
+  }
+
+  checkValidity() {
+    if (!this.willValidate) {
+      return true;
+    }
+    const { validity } = this;
+    if (!validity.valid) {
+      this._dispatch(Event.createImpl(["invalid", { cancelable: false } ]));
+    }
+    return validity.valid;
+  }
+
+  setCustomValidity(message) {
+    this._customValidityMessage = message;
```

Comment:
1. This function appears in multiple places. Would it be possible to refactor this function (and possibly other related methods) into a dedicated mixin? Same for the implementations of `willValidate` and `customError` state methods in the `ValidityState` creation, which are the same for all elements. Only the impl method needs to be factored out, no need to divide the IDL file as well. Check out `git grep '^mixin('`.
2. Nit: `_customValidityErrorMessage` per [spec](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#custom-validity-error-message)
3. Make sure to initialize this property to the empty string in the constructor, per spec. If this method is factored out into a dedicated mixin, create a function called `_initFormControl` in which this property is initialized, and call `this._initFormControl()` in each element's constructor. This should also allow you to get rid of the `this._customValidityMessage != null` check in `customError`.